### PR TITLE
Re-land "add IntoFuture trait and support for await"

### LINKED ISSUE #67982: Re-land "add IntoFuture trait and support for await"
#65244 was reverted in https://github.com/rust-lang/rust/pull/67768 for performance reasons (see https://github.com/rust-lang/rust/issues/67706). This issue tracks re-landing the support in a more performant way.

Tracking issue for `IntoFuture` trait:  https://github.com/rust-lang/rust/issues/67644

cc @seanmonstar @wesleywiser

### PR DESCRIPTION
**Pull request by tmandry**
Testing the code from #65244 to see if the performance regressions are still there. #68606 and #68672 made perf optimizations that might interact with this change.

If this lands, fixes #67982.

cc @seanmonstar @jonas-schievink 
r? @cramertj

### CODE REVIEW

**Comment by tmandry** [comment]
@bors try
@rust-timer queue

**Comment by Arnavion** [comment]
In case you haven't already, please also ensure https://github.com/rust-lang/rust/issues/67757 doesn't get reintroduced. Thanks.

**Comment by tmandry** [comment]
@Arnavion Thanks for pointing to that, I'd indeed forgotten about the type length limit problem. I'm not quite sure what to do about it. We discussed this during the original PR ([around here](https://github.com/rust-lang/rust/pull/65244#issuecomment-545119360)) and thought we had a fix for it in https://github.com/rust-lang/rust/pull/65743, but that seems to have only been a partial fix. cc @eddyb 

The perf run seems to be done; [results](https://perf.rust-lang.org/compare.html?start=8417d68de5e063426ab6bb7f383df6117d1beeed&end=976ac3b5cd579bc2040b7473df7d885790a47082) are not great for async benchmarks, but not a total disaster either. (This is far better than the [original regression](https://perf.rust-lang.org/compare.html?start=74c4e6a981d3150db8444c8d250e50bbe6b93b6b&end=3a087ad3a924be12343bb035bf9b63ed81f650bf&stat=instructions:u) which led us to revert the change.) While it might be possible to land the change with this regression, I'd prefer to dig into it some more first.

So, in summary, I think we need a fix to #67757, and to dig into the remaining performance regression. If anyone wants to take on either of these (or both), please feel free!

File: src/librustc_ast_lowering/expr.rs [hunk 1]
```diff
@@ -521,7 +521,7 @@ impl<'hir> LoweringContext<'_, 'hir> {
 
     /// Desugar `<expr>.await` into:
     /// ```rust
-    /// match <expr> {
+    /// match ::std::future::IntoFuture::into_future(<expr>) {
```
Comment (eddyb) [inline src/librustc_ast_lowering/expr.rs]: I don't see how #67757 is possible, the only HIR difference here should be handled by #65743.

**Comment by Aaron1011** [comment]
I think https://github.com/rust-lang/rust/pull/69218 should help improve performance here - this PR appears to introduce many additional predicates during typeck, which end up getting stalled for several `select` iterations.

**Comment by joelpalmer** [comment]
Triaged

**Comment by Dylan-DPC** [comment]
@tmandry any updates?

**Comment by Dylan-DPC** [comment]
Closing this due to inactivity
